docs/configuration.md
# Configuration

## Map Marker Icons

The CB-Frontend library comes with a map marker icon generation utility
which makes it relatively easy to configure custom icons for your map.

These marker icon generation utilities are called renderers. You can define multiple renderers
in your configuration. If a renderer cannot yield an icon for a common, the next renderer in
your configuration is used as a fallback.

For instance the _thumbnail_ renderer can only yield an icon if an image has been assigned
to a common. If the common has no image, the _thumbnail_ renderer does not yield an icon
and the next renderer in the configuration is used.

See the examples and the reference below.

### Examples

#### Category Icons and Thumbnails

This configuration example

1. first tries to match certain categories
2. if no category has a match, the thumbnail is used
3. if the common has no thumbnail, a static color is used

```json
{
  "map": {
    "markerIcon": {
      "renderers": [
        {
          "type": "category",
          "match": [
            {
              "categories": [6, 8],
              "renderers": [{ "type": "image", "url": "/assets/kasten-elektrisch.png" }]
            },
            {
              "categories": [6],
              "renderers": [{ "type": "image", "url": "/assets/elektrisch.png" }]
            },
            { "categories": [8], "renderers": [{ "type": "image", "url": "/assets/kasten.png" }] },
            {
              "categories": [12],
              "renderers": [{ "type": "image", "url": "/assets/3-raeder.png" }]
            },
            { "categories": [16], "renderers": [{ "type": "color", "color": "teal" }] }
          ]
        },
        { "type": "thumbnail" },
        { "type": "color", "color": "hsl(20 60% 80%)" }
      ]
    }
  }
}
```

#### Traditional map marker

This configuration example simply uses the standard map marker icon we’re all used to.

```json
{
  "map": {
    "markerIcon": {
      "renderers": [{ "type": "traditional-icon" }]
    }
  }
}
```

#### Custom map marker icon

This configuration example uses a custom map marker icon.

```json
{
  "map": {
    "markerIcon": {
      "renderers": [
        {
          "type": "icon",
          "url": "/assets/my-custom-icon.png",
          "width": 25,
          "height": 40,
          "anchor": { "x": 0.5, "y": 1 }
        }
      ]
    }
  }
}
```

### Renderer Reference

#### Color

Renders a static color inside a map marker.

```json
{ "type": "color", "color": "#cc0000" }
```

#### Image

Renders a static image inside a map marker.

```json
{ "type": "image", "url": "/assets/my-image.jpg" }
```

#### Icon

Renders a custom icon.

```json
{
  "type": "icon",
  "url": "/assets/my-icon.png",
  "width": 25,
  "height": 40,
  "anchor": { "x": 0.5, "y": 1 }
}
```

#### Traditional Icon

Renders a traditional map marker icon shipped with the CB-Frontend library.

```json
{ "type": "traditional-icon" }
```

#### Thumbnail

Renders the common’s thumbnail image inside a map marker.
Does not work as user marker icon.

```json
{ "type": "thumbnail" }
```

#### Category

Checks if all the category ids match for the given common and renders the appropriate marker.
The first match wins. Does not work as user marker icon.

```json
{
  "type": "category",
  "match": [
    { "categories": [1, 2, 3], "renderers": [{ "type": "color", "color": "green" }] },
    { "categories": [2], "renderers": [{ "type": "color", "color": "hsl(90 60% 80%)" }] }
  ]
}
```

### Icon Envelope

The _color_, _image_, and _thumbnail_ renderer all use the same "envelope" icon.
This envelope is also configurable. All the mentioned renderer types, as well as the
_category_ renderer, and the individual category match instances support the `wrap`
attribute that takes a configuration object.

This object looks like this (all keys are optional):

<!-- prettier-ignore-start -->
```json5
{
  // The color of the envelope itself.
  "fill": "#fff",
  // A number that scales the icon proportionally.
  "scale": 0.5,
  // An SVG image string is used for creating icons.
  // This must NOT be a path, but the actual SVG string.
  // See: /src/assets/map-marker-template.svg
  "template": "...",
  // The background color of the circle that embeds the content.
  // You might want to change this, if you want to use an image renderer
  // with a transparent image background.
  "embedFill": "yellow",
  // A label that is displayed at the center of the circle.
  // Only relevant for cluster icon.
  "embedLabel": "...",
  // The text color of the label at the center of the circle.
  // Only relevant for cluster icon.
  "embedLabelStroke": "red"
}
```
<!-- prettier-ignore-end -->

You can also modify the envelope globally by using the `wrapDefaults` property.

In the following example we half the size of the envelope and change its color,
but increase the size by a factor of 5, if a thumbnail is rendered.

```json
{
  "map": {
    "markerIcon": {
      "wrapDefaults": { "scale": 0.5, "fill": "#eaeaea" },
      "renderers": [
        { "type": "thumbnail", "wrap": { "scale": 5 } },
        { "type": "color", "color": "hsl(20 60% 80%)" }
      ]
    }
  }
}
```

### User location icon

The `map.markerIcon` property is only used for the commons on the map.
If you want to change the icon for the user’s location on the map you
must set the `map.userMarkerIcon` property instead. All renderer types are supported
except for _thumbnail_ and _category_.

### Cluster icon

If you have enabled icon clustering by setting the `map.cluster.radiusPixels` option
a cluster icon will be shown if commons are too close to each other.

You can modify the default icon by setting the `map.cluster.markerIcon` option.
Only the `color` renderer type is supported, but the icon envelope can be modified as well.

Example:

```json
{
  "map": {
    "cluster": {
      "radiusPixels": 60,
      "markerIcon": {
        "renderers": [{ "type": "color", "color": "crimson", "labelColor": "white" }]
      }
    }
  }
}
```


src/commons-search/fixtures/settings.ts
import { addDays } from 'date-fns';
import { toDateString } from '../../util';
import { CommonMarkerIconConfig, MarkerIconConfig, ExtendedLegacyMapConfiguration } from '../types';
import LightningIcon from './lightning.svg?url';
import HandIcon from './hand.svg?url';

const markerIcon: CommonMarkerIconConfig = {
  renderers: [
    {
      type: 'category',
      match: [
        {
          categories: [29],
          renderers: { type: 'image', url: LightningIcon },
        },
        {
          categories: [31],
          renderers: {
            type: 'color',
            color: 'hsl(230 60% 80%)',
          },
        },
      ],
    },
    { type: 'thumbnail' },
  ],
  wrapDefaults: { scale: 0.75 },
};

const userMarkerIcon: MarkerIconConfig = {
  renderers: [{ type: 'image', url: HandIcon }],
};

export default {
  version: 1,
  dataSource: { type: 'fixtures' },
  custom_marker_icon: null,
  item_draft_marker_icon: null,
  preferred_status_marker_icon: 'publish',
  filter_cb_item_categories: {
    'g1610805575766-21823': {
      name: '',
      elements: [
        {
          cat_id: 27,
          markup: '2 Räder',
        },
        {
          cat_id: 28,
          markup: '3 Räder',
        },
        {
          cat_id: 35,
          markup: 'Gespann',
        },
      ],
    },
    'g1610805862646-605141': {
      name: 'Antrieb',
      elements: [
        {
          cat_id: 29,
          markup: 'Elektro',
        },
        {
          cat_id: 30,
          markup: 'Muskelkraft',
        },
      ],
    },
    'g1610805870910-325344': {
      name: '',
      elements: [
        {
          cat_id: 31,
          markup: 'Kiste mit Schloss',
        },
        {
          cat_id: 32,
          markup: 'Regenverdeck',
        },
      ],
    },
    'g1610805880437-723630': {
      name: '',
      elements: [
        {
          cat_id: 34,
          markup: 'Kindertransport',
        },
        {
          cat_id: 33,
          markup: 'Rollstuhl',
        },
      ],
    },
  },
  filter_availability: {
    date_min: toDateString(new Date()),
    date_max: toDateString(addDays(new Date(), 31)),
    day_count_max: '20',
  },
  cb_map_id: 7561,
  locale: 'de-DE',
  asset_path: 'https://stage3.commons-booking.org/wp-content/plugins/commonsbooking/assets/map/',
  base_map: 1,
  show_scale: false,
  zoom_min: 9,
  zoom_max: 19,
  scrollWheelZoom: true,
  zoom_start: 11,
  lat_start: 50.953128,
  lon_start: 6.945275,
  marker_map_bounds_initial: false,
  marker_map_bounds_filter: false,
  max_cluster_radius: 80,
  marker_tooltip_permanent: false,
  show_location_contact: false,
  show_location_opening_hours: false,
  show_item_availability: true,
  show_location_distance_filter: true,
  label_location_distance_filter: '',
  show_item_availability_filter: false,
  label_item_availability_filter: 'Verfügbarkeit',
  label_item_category_filter: 'Merkmale',
  map: {
    markerIcon,
    userMarkerIcon,
  },
  geocode: {
    region: { countryCodes: ['de'] },
  },
} as ExtendedLegacyMapConfiguration;
